Weblogic exception on agent startup : Class not Found weblogic.security.SSL.jsseadapter.JaSSLContextImpl

Discovered with `1.28.4` agent + Weblogic 12c `12.2.1.3.0`, very likely to impact other versions as well.

## Description

The following stack trace is visible in Weblogic standard output when the agent starts

```
2022-01-20 17:23:11,235 [elastic-apm-server-healthcheck] ERROR co.elastic.apm.agent.util.ExecutorUtils - [Security:090461]Class not Found weblogic.security.SSL.jsseadapter.JaSSLContextImpl
java.lang.IllegalArgumentException: [Security:090461]Class not Found weblogic.security.SSL.jsseadapter.JaSSLContextImpl
        at weblogic.security.utils.SSLSetup.getDelegateClass(SSLSetup.java:146) ~[?:?]
        at weblogic.security.utils.SSLSetup.getSSLDelegateInstance(SSLSetup.java:103) ~[?:?]
        at weblogic.security.utils.SSLContextWrapper.<init>(SSLContextWrapper.java:82) ~[?:?]
        at weblogic.security.utils.SSLContextWrapper.getInstance(SSLContextWrapper.java:68) ~[?:?]
        at weblogic.security.utils.SSLSetup.getSSLContext(SSLSetup.java:368) ~[?:?]
        at weblogic.security.utils.SSLSetup.getSSLContext(SSLSetup.java:360) ~[?:?]
        at weblogic.security.SSL.SSLClientInfo.getSSLSocketFactory(SSLClientInfo.java:111) ~[?:?]
        at weblogic.security.SSL.SSLSocketFactory.setSSLClientInfo(SSLSocketFactory.java:243) ~[?:?]
        at weblogic.security.SSL.SSLSocketFactory.<init>(SSLSocketFactory.java:46) ~[?:?]
        at weblogic.security.SSL.SSLSocketFactory.getInstance(SSLSocketFactory.java:79) ~[?:?]
        at weblogic.net.http.HttpsClient.getInternalSocketFactory(HttpsClient.java:263) ~[?:?]
        at weblogic.net.http.HttpsClient.openServer(HttpsClient.java:272) ~[?:?]
        at weblogic.net.http.HttpsClient.openServer(HttpsClient.java:362) ~[?:?]
        at weblogic.net.http.HttpsClient.New(HttpsClient.java:565) ~[?:?]
        at weblogic.net.http.HttpsClient.New(HttpsClient.java:546) ~[?:?]
        at weblogic.net.http.HttpsURLConnection.connect(HttpsURLConnection.java:235) ~[?:?]
        at weblogic.net.http.HttpURLConnection.getInputStream(HttpURLConnection.java:685) ~[?:?]
        at weblogic.net.http.SOAPHttpsURLConnection.getInputStream(SOAPHttpsURLConnection.java:42) ~[?:?]
        at co.elastic.apm.agent.report.HttpUtils.consumeAndClose(HttpUtils.java:62) ~[elastic-apm-agent-1.28.4.jar:?]
        at co.elastic.apm.agent.report.ApmServerClient.executeForAllUrls(ApmServerClient.java:280) ~[elastic-apm-agent-1.28.4.jar:?]
        at co.elastic.apm.agent.report.ApmServerHealthChecker.call(ApmServerHealthChecker.java:65) ~[elastic-apm-agent-1.28.4.jar:?]
        at co.elastic.apm.agent.report.ApmServerHealthChecker.call(ApmServerHealthChecker.java:42) ~[elastic-apm-agent-1.28.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_311]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_311]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_311]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_311]
Caused by: java.lang.ClassNotFoundException: weblogic/security/SSL/jsseadapter/JaSSLContextImpl
        at java.lang.Class.forName0(Native Method) ~[?:1.8.0_311]
        at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_311]
        at weblogic.security.utils.SSLSetup.loadDelegateClass(SSLSetup.java:161) ~[?:?]
        at weblogic.security.utils.SSLSetup.getDelegateClass(SSLSetup.java:133) ~[?:?]
        ... 25 more
2022-01-20 17:23:12,102 [elastic-apm-agent-initialization] INFO  co.elastic.apm.agent.impl.ElasticApmTracer - Tracer switched to RUNNING state
```

## Impact

~~Low, does not prevent the agent to send data to APM server~~
Could be quite high, as it works "by accident" of the agent being fast on JVM startup, and it can prevent the agent to communicate with APM server. Would also happen systematically when using runtime attach.

--- a/CHANGELOG.asciidoc
+++ b/CHANGELOG.asciidoc
@@ -34,6 +34,7 @@ Use subheadings with the "=====" level for adding notes for unreleased changes:
 [float]
 ===== Bug fixes
 * Prevent NPE in OpenTelemetry metrics bridge in case of asynchronous agent start - {pull}3880[#3880]
+* Fix random Weblogic ClassNotFoundException related to thread context classloader - {pull}3870[#3870]
 
 [[release-notes-1.x]]
 === Java Agent version 1.x

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/report/AbstractIntakeApiHandler.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/report/AbstractIntakeApiHandler.java
@@ -22,6 +22,7 @@ import co.elastic.apm.agent.report.serialize.DslJsonSerializer;
 import co.elastic.apm.agent.report.serialize.SerializationConstants;
 import co.elastic.apm.agent.sdk.logging.Logger;
 import co.elastic.apm.agent.sdk.logging.LoggerFactory;
+import co.elastic.apm.agent.util.UrlConnectionUtils;
 import org.stagemonitor.util.IOUtils;
 
 import javax.annotation.Nullable;
@@ -93,63 +94,64 @@ public class AbstractIntakeApiHandler {
     protected HttpURLConnection startRequest(String endpoint) throws Exception {
         payloadSerializer.blockUntilReady();
         final HttpURLConnection connection = apmServerClient.startRequest(endpoint);
-        if (connection != null) {
-            boolean useCompression = !isLocalhost(connection);
-            try {
-                if (logger.isDebugEnabled()) {
-                    logger.debug("Starting new request to {}", connection.getURL());
-                }
-                connection.setRequestMethod("POST");
-                connection.setDoOutput(true);
-                connection.setChunkedStreamingMode(SerializationConstants.BUFFER_SIZE);
-                if (useCompression) {
-                    connection.setRequestProperty("Content-Encoding", "deflate");
-                }
-                connection.setRequestProperty("Content-Type", "application/x-ndjson");
-                connection.setUseCaches(false);
-                connection.connect();
-                countingOs = new CountingOutputStream(connection.getOutputStream());
-                if (useCompression) {
-                    os = new DeflaterOutputStream(countingOs, deflater, true);
-                } else {
-                    os = countingOs;
-                }
-                payloadSerializer.setOutputStream(os);
-                payloadSerializer.appendMetaDataNdJsonToStream();
-                payloadSerializer.flushToOutputStream();
-                requestStartedNanos = System.nanoTime();
-            } catch (IOException e) {
-                try {
-                    logger.error("Error trying to connect to APM Server at {}. Although not necessarily related to SSL, some related SSL " +
-                        "configurations corresponding the current connection are logged at INFO level.", connection.getURL());
-                    if (logger.isInfoEnabled() && connection instanceof HttpsURLConnection) {
-                        HttpsURLConnection httpsURLConnection = (HttpsURLConnection) connection;
-                        try {
-                            logger.info("Cipher suite used for this connection: {}", httpsURLConnection.getCipherSuite());
-                        } catch (Exception e1) {
-                            SSLSocketFactory sslSocketFactory = httpsURLConnection.getSSLSocketFactory();
-                            logger.info("Default cipher suites: {}", Arrays.toString(sslSocketFactory.getDefaultCipherSuites()));
-                            logger.info("Supported cipher suites: {}", Arrays.toString(sslSocketFactory.getSupportedCipherSuites()));
-                        }
-                        try {
-                            logger.info("APM Server certificates: {}", Arrays.toString(httpsURLConnection.getServerCertificates()));
-                        } catch (Exception e1) {
-                            // ignore - invalid
-                        }
-                        try {
-                            logger.info("Local certificates: {}", Arrays.toString(httpsURLConnection.getLocalCertificates()));
-                        } catch (Exception e1) {
-                            // ignore - invalid
-                        }
-                    }
-                } finally {
-                    closeAndSuppressErrors(connection);
-                }
-                throw e;
-            } catch (Throwable t) {
-                closeAndSuppressErrors(connection);
-                throw t;
+        if (connection == null) {
+            return null;
+        }
+        try (UrlConnectionUtils.ContextClassloaderScope clScope = UrlConnectionUtils.withContextClassloaderOf(connection)){
+            if (logger.isDebugEnabled()) {
+                logger.debug("Starting new request to {}", connection.getURL());
             }
+            boolean useCompression = !isLocalhost(connection);
+            connection.setRequestMethod("POST");
+            connection.setDoOutput(true);
+            connection.setChunkedStreamingMode(SerializationConstants.BUFFER_SIZE);
+            if (useCompression) {
+                connection.setRequestProperty("Content-Encoding", "deflate");
+            }
+            connection.setRequestProperty("Content-Type", "application/x-ndjson");
+            connection.setUseCaches(false);
+            connection.connect();
+            countingOs = new CountingOutputStream(connection.getOutputStream()); // TODO : here
+            if (useCompression) {
+                os = new DeflaterOutputStream(countingOs, deflater, true);
+            } else {
+                os = countingOs;
+            }
+            payloadSerializer.setOutputStream(os);
+            payloadSerializer.appendMetaDataNdJsonToStream();
+            payloadSerializer.flushToOutputStream();
+            requestStartedNanos = System.nanoTime();
+        } catch (IOException e) {
+            try {
+                logger.error("Error trying to connect to APM Server at {}. Although not necessarily related to SSL, some related SSL " +
+                    "configurations corresponding the current connection are logged at INFO level.", connection.getURL());
+                if (logger.isInfoEnabled() && connection instanceof HttpsURLConnection) {
+                    HttpsURLConnection httpsURLConnection = (HttpsURLConnection) connection;
+                    try {
+                        logger.info("Cipher suite used for this connection: {}", httpsURLConnection.getCipherSuite());
+                    } catch (Exception e1) {
+                        SSLSocketFactory sslSocketFactory = httpsURLConnection.getSSLSocketFactory();
+                        logger.info("Default cipher suites: {}", Arrays.toString(sslSocketFactory.getDefaultCipherSuites()));
+                        logger.info("Supported cipher suites: {}", Arrays.toString(sslSocketFactory.getSupportedCipherSuites()));
+                    }
+                    try {
+                        logger.info("APM Server certificates: {}", Arrays.toString(httpsURLConnection.getServerCertificates()));
+                    } catch (Exception e1) {
+                        // ignore - invalid
+                    }
+                    try {
+                        logger.info("Local certificates: {}", Arrays.toString(httpsURLConnection.getLocalCertificates()));
+                    } catch (Exception e1) {
+                        // ignore - invalid
+                    }
+                }
+            } finally {
+                closeAndSuppressErrors(connection);
+            }
+            throw e;
+        } catch (Throwable t) {
+            closeAndSuppressErrors(connection);
+            throw t;
         }
         return connection;
     }
@@ -188,7 +190,10 @@ public class AbstractIntakeApiHandler {
     }
 
     private void endRequest(boolean isFailed) {
-        if (connection != null) {
+        if (connection == null) {
+            return;
+        }
+        try (UrlConnectionUtils.ContextClassloaderScope clScope = UrlConnectionUtils.withContextClassloaderOf(connection)) {
             long writtenBytes = countingOs != null ? countingOs.getCount() : 0L;
             try {
                 payloadSerializer.fullFlush();

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java
@@ -265,8 +265,10 @@ public class ApmServerClient {
         Exception previousException = null;
         for (URL serverUrl : prioritizedUrlList) {
             HttpURLConnection connection = null;
+            UrlConnectionUtils.ContextClassloaderScope clScope = null;
             try {
                 connection = startRequestToUrl(appendPath(serverUrl, path));
+                clScope = UrlConnectionUtils.withContextClassloaderOf(connection);
                 return connectionHandler.withConnection(connection);
             } catch (Exception e) {
                 expectedErrorCount = incrementAndGetErrorCount(expectedErrorCount);
@@ -277,6 +279,9 @@ public class ApmServerClient {
                 previousException = e;
             } finally {
                 HttpUtils.consumeAndClose(connection);
+                if (clScope != null) {
+                    clScope.close();
+                }
             }
         }
         if (previousException == null) {
@@ -290,13 +295,19 @@ public class ApmServerClient {
         List<T> results = new ArrayList<>(serverUrls.size());
         for (URL serverUrl : serverUrls) {
             HttpURLConnection connection = null;
+            UrlConnectionUtils.ContextClassloaderScope clScope = null;
             try {
                 connection = startRequestToUrl(appendPath(serverUrl, path));
+                clScope = UrlConnectionUtils.withContextClassloaderOf(connection);
                 results.add(connectionHandler.withConnection(connection));
             } catch (Exception e) {
                 logger.debug("Exception while interacting with APM Server", e);
             } finally {
                 HttpUtils.consumeAndClose(connection);
+                if (clScope != null) {
+                    clScope.close();
+                }
+
             }
         }
         return results;
@@ -436,6 +447,7 @@ public class ApmServerClient {
     /**
      * Escapes the provided string from characters that are disallowed within HTTP header comments.
      * See spec- https://httpwg.org/specs/rfc7230.html#field.components
+     *
      * @param headerFieldComment HTTP header comment value to be escaped
      * @return the escaped header comment
      */

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/report/PartialTransactionReporter.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/report/PartialTransactionReporter.java
@@ -26,6 +26,7 @@ import co.elastic.apm.agent.report.serialize.SerializationConstants;
 import co.elastic.apm.agent.sdk.logging.Logger;
 import co.elastic.apm.agent.sdk.logging.LoggerFactory;
 import co.elastic.apm.agent.tracer.pooling.Allocator;
+import co.elastic.apm.agent.util.UrlConnectionUtils;
 
 import java.net.HttpURLConnection;
 
@@ -67,27 +68,30 @@ class PartialTransactionReporter {
                 logger.debug("Cannot report partial transaction because server url is not configured");
                 return;
             }
-            connection.setRequestMethod("POST");
-            connection.setDoOutput(true);
-            connection.setChunkedStreamingMode(SerializationConstants.BUFFER_SIZE);
-            connection.setRequestProperty("Content-Type", "application/vnd.elastic.apm.transaction+ndjson");
-            connection.setRequestProperty("x-elastic-aws-request-id", requestId);
-            connection.setUseCaches(false);
-            connection.connect();
 
-            DslJsonSerializer.Writer writer = writerPool.createInstance();
-            try {
-                writer.setOutputStream(connection.getOutputStream());
-                writer.blockUntilReady(); //should actually not block on AWS Lambda, as metadata is available immediately
-                writer.appendMetaDataNdJsonToStream();
-                writer.serializeTransactionNdJson(transaction);
-                writer.fullFlush();
-            } finally {
-                writerPool.recycle(writer);
+            try (UrlConnectionUtils.ContextClassloaderScope clScope = UrlConnectionUtils.withContextClassloaderOf(connection)) {
+                connection.setRequestMethod("POST");
+                connection.setDoOutput(true);
+                connection.setChunkedStreamingMode(SerializationConstants.BUFFER_SIZE);
+                connection.setRequestProperty("Content-Type", "application/vnd.elastic.apm.transaction+ndjson");
+                connection.setRequestProperty("x-elastic-aws-request-id", requestId);
+                connection.setUseCaches(false);
+                connection.connect();
+
+                DslJsonSerializer.Writer writer = writerPool.createInstance();
+                try {
+                    writer.setOutputStream(connection.getOutputStream());
+                    writer.blockUntilReady(); //should actually not block on AWS Lambda, as metadata is available immediately
+                    writer.appendMetaDataNdJsonToStream();
+                    writer.serializeTransactionNdJson(transaction);
+                    writer.fullFlush();
+                } finally {
+                    writerPool.recycle(writer);
+                }
+
+                handleResponse(connection);
+                connection.disconnect();
             }
-
-            handleResponse(connection);
-            connection.disconnect();
 
         } catch (Exception e) {
             logger.error("Failed to report partial transaction {}", transaction, e);

--- a/apm-agent-core/src/main/java/co/elastic/apm/agent/util/UrlConnectionUtils.java
+++ b/apm-agent-core/src/main/java/co/elastic/apm/agent/util/UrlConnectionUtils.java
@@ -23,7 +23,9 @@ import co.elastic.apm.agent.sdk.internal.util.PrivilegedActionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.annotation.Nullable;
 import java.io.IOException;
+import java.net.HttpURLConnection;
 import java.net.Proxy;
 import java.net.ProxySelector;
 import java.net.URISyntaxException;
@@ -85,6 +87,50 @@ public class UrlConnectionUtils {
             } catch (URISyntaxException e) {
                 logger.debug("Failed to read and debug-print proxy settings for {}", url, e);
             }
+        }
+    }
+
+    public static ContextClassloaderScope withContextClassloaderOf(HttpURLConnection connection) {
+        Class<?> type = connection.getClass();
+        // so far, only weblogic is known to use context CL to load parts of the SSL/TLS communication
+        // see https://github.com/elastic/apm-agent-java/issues/2409 for background on why this is needed
+        boolean override = type.getCanonicalName().startsWith("weblogic.");
+        return withContextClassloader(type.getClassLoader(), override);
+    }
+
+    // package private for testing
+    static ContextClassloaderScope withContextClassloader(@Nullable ClassLoader cl, boolean override) {
+        if (!override) {
+            return ContextClassloaderScope.NOOP;
+        }
+        return new ContextClassloaderScope(cl, false);
+    }
+
+    public static class ContextClassloaderScope implements AutoCloseable {
+
+        private static final ContextClassloaderScope NOOP = new ContextClassloaderScope(null, true);
+
+        private final boolean noop;
+        @Nullable
+        private final ClassLoader previous;
+
+        private ContextClassloaderScope(@Nullable ClassLoader classLoader, boolean noop) {
+            this.noop = noop;
+            if (noop) {
+                this.previous = null;
+                return;
+            }
+            this.previous = Thread.currentThread().getContextClassLoader();
+            Thread.currentThread().setContextClassLoader(classLoader);
+        }
+
+        @Override
+        public void close() {
+            if (noop) {
+                return;
+            }
+            Thread.currentThread().setContextClassLoader(previous);
+
         }
     }
 }

--- a/apm-agent-core/src/test/java/co/elastic/apm/agent/util/UrlConnectionUtilsTest.java
+++ b/apm-agent-core/src/test/java/co/elastic/apm/agent/util/UrlConnectionUtilsTest.java
@@ -1,0 +1,80 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.util;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import javax.annotation.Nullable;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class UrlConnectionUtilsTest {
+
+    @BeforeEach
+    void beforeEach() {
+        Thread.currentThread().setContextClassLoader(null);
+    }
+
+    @AfterEach
+    void afterEach() {
+        Thread.currentThread().setContextClassLoader(null);
+    }
+
+    @Test
+    void threadContextClassLoader_noop() {
+        ClassLoader cl = UrlConnectionUtils.class.getClassLoader();
+        testContextClassLoader(null, null, false);
+        testContextClassLoader(cl, cl, false);
+    }
+
+    @Test
+    void threadContextClassLoader_override() {
+        ClassLoader cl = UrlConnectionUtils.class.getClassLoader();
+        testContextClassLoader(null, cl, true);
+        testContextClassLoader(cl, cl, true);
+
+        DummyClassLoader overrideCl = new DummyClassLoader();
+        testContextClassLoader(null, overrideCl, true);
+        testContextClassLoader(cl, overrideCl, true);
+    }
+
+    private void testContextClassLoader(@Nullable ClassLoader initialClassLoader, @Nullable ClassLoader scopeClassLoader, boolean override) {
+        Thread.currentThread().setContextClassLoader(initialClassLoader);
+        checkContextClassLoader(initialClassLoader);
+        try (UrlConnectionUtils.ContextClassloaderScope scope = UrlConnectionUtils.withContextClassloader(scopeClassLoader, override)) {
+            checkContextClassLoader(scopeClassLoader);
+        }
+        checkContextClassLoader(initialClassLoader);
+    }
+
+    private void checkContextClassLoader(@Nullable ClassLoader expected) {
+        ClassLoader threadContextCl = Thread.currentThread().getContextClassLoader();
+        if (expected == null) {
+            assertThat(threadContextCl).isNull();
+        } else {
+            assertThat(threadContextCl).isSameAs(expected);
+        }
+    }
+
+    private static class DummyClassLoader extends ClassLoader {
+
+    }
+}

--- a/docs/troubleshooting.asciidoc
+++ b/docs/troubleshooting.asciidoc
@@ -288,6 +288,10 @@ further information about how to download this file can be found <<setup-javaage
 `java.net.SocketException: Broken pipe` - one option is that client authentication fails. Check out
 <<ssl-client-authentication>>.
 
+With Oracle Weblogic application server wildcard TLS certificates are not allowed by default, when this
+happens a message error like this is visible in agent logs: `Hostname verification failed: HostnameVerifier=weblogic.security.utils.SSLWLSHostnameVerifier`.
+Disabling this extra check can be done with `-Dweblogic.security.SSL.ignoreHostnameVerification=true`.
+
 For other SSL/TLS related problems, - check out
 https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html#Troubleshooting[the JSSE
 troubleshooting section].

